asmcomp/power/emit.mlp

```diff
@@ -1041,14 +1062,24 @@ let fundecl fundecl =
   end;
   emit_debug_info fundecl.fun_dbg;
   cfi_startproc();
-  (* On this target, there is at most one "out of line" code block per
-     function: a single "call GC" point.  It comes immediately after the
-     function's body. *)
-  BR.relax fundecl.fun_body ~max_out_of_line_code_offset:0;
+  let num_call_gc = num_call_gc fundecl.fun_body in
+  let max_out_of_line_code_offset = max (num_call_gc - 1) 0 in
+  BR.relax fundecl.fun_body ~max_out_of_line_code_offset;
   emit_all fundecl.fun_body;
   (* Emit the glue code to call the GC *)
-  if !call_gc_label > 0 then begin
-    `{emit_label !call_gc_label}:\n`;
+  assert (IntMap.cardinal !call_gc_labels = num_call_gc);
+  if num_call_gc > 0 then begin
+    (* Replace sizes by deltas with next size *)
+    let rec delta_encode = function
+      | (sz1, lbl1) :: ((sz2, _) :: _ as l) ->
+           (sz1 - sz2, lbl1) :: delta_encode l
+      | ([] | [(_,_)]) as l -> l in
+    let delta_lbl_list =
+      delta_encode (List.rev (IntMap.bindings !call_gc_labels)) in
```

Comment:
This encoding technique uses `n` instructions for declarations of `n` different sizes (one addition per label), instead of `2n+1` (one addition and a jump), at the cost of making each call slide through up to `n` different additions, The innocuous `List.rev` call here will have a noticeable impact on performance, as it guarantees that the largest calls are at the beginning of the addition slider, taking more time, and the (much more common) shortest calls are at the end, taking less time.

I think it could be nice to comment on the explicit choice of order here.

(Would it make sense to handle the most common allocations, sizes 1-5, with an extra jump, and use the slider approach only for larger sizes? (Separating the two is easy with `IntMap.split 6`, but of course this requires to implement the two logics.). I suppose that you didn't do this because there are fewer users of the architecture, so "simple" is better than "optimized".)

Reply:
Generally speaking, I have the impression (not yet backed by numbers) that the number N of different allocation sizes per function is small, even when the function is quite large, hence the cost of going through at most N additions is very small.  Also, remember that few allocation points call the GC, signal handling notwithstanding.

The main reason for sorting allocation sizes in decreasing order was to make the generated code more legible: all deltas are positive, hence we can "see" the allocation pointer being increased until it reaches its initial value.  Any other order would be semantically correct but look odd, with the allocation pointer being decreased sometimes, increased some other times.  

The fact that smaller allocations go through a shorter sequence of adds was not planned, but is a nice touch :-)  Thank you for pointing it out.

I promise I'll put a comment.

Reply:
The comment was added.